src/spi.rs

```diff
@@ -0,0 +1,180 @@
+
+
+use std::io::{self, Read};
+use std::sync::{Arc, Mutex};
+
+use hal::blocking::spi;
+
+use ::error::MockError;
+
+/// SPI Transaction mode
+#[derive(Clone, Debug, PartialEq)]
+pub enum Mode {
+    None,
+    Write,
+    Transfer,
+}
+
+/// SPI transaction type
+#[derive(Clone, Debug, PartialEq)]
+pub struct Transaction {
+    expected_mode: Mode,
+    expected_data: Vec<u8>,
+    response: Vec<u8>,
+}
+
+impl Transaction {
+    /// Create a write transaction
+    pub fn write(expected: Vec<u8>) -> Transaction {
+        Transaction{
+            expected_mode: Mode::Write,
+            expected_data: expected,
+            response: Vec::new(),
+        }
+    }
+
+    /// Create a transfer transaction
+    pub fn transfer(expected: Vec<u8>, response: Vec<u8>) -> Transaction {
+        Transaction{
+            expected_mode: Mode::Transfer,
+            expected_data: expected,
+            response,
+        }
+    }
+}
+
+/// Mock SPI implementation
+pub struct SPIMock {
+    expected: Arc<Mutex<Vec<Transaction>>>,
+}
+
+impl SPIMock {
+    /// Create a new mock SPI interface
+    pub fn new() -> Self {
+        SPIMock { expected: Arc::new(Mutex::new(Vec::new())) }
+    }
+
+    /// Set expectations on the SPI interface
+    /// This is a list of SPI transactions to be executed
+    pub fn expect(&mut self, expected: Vec<Transaction>) {
+        let mut e = self.expected.lock().unwrap();
+        assert_eq!(e.len(), 0);
```

Comment:
I'm not sure if setting expectations should implicitly check previous state... I would expect the `expect` method to simply overwrite the previous expectations.

Maybe an `add_expect(&mut self, expected: Transaction)` might be interesting as an `append-this-to-current-expectations` kind of method?